File: searchForSubstring.js
//zoho problem
//strore a string in a two dimensional array
//find the given word in that two dimensional array both from left to right and from top to bottom

const str = "WELCOMETOZOHOCORPORATION";
let word = "ocr";
let strArray = [];
let temp = [];
for (let i = 0; i < str.length; i++) {
  temp.push(str[i]);
  if ((i + 1) % 5 == 0 || i == str.length - 1) {
    strArray.push(temp);
    temp = [];
  }
}
let indexes = "";
let count = 0;
for (let i = 0; i < strArray.length; i++) {
  for (let j = 0; j < strArray[i].length; j++) {
    //for left to right
    if (word[0].toUpperCase() == strArray[i][j]) {
      for (let k = 0; k < word.length; k++) {
        if (
          strArray[i][j + k] &&
          word[count].toUpperCase() == strArray[i][j + k]
        ) {
          count++;
          if (count == word.length) {
            indexes = `<${i},${j}>\n<${i},${j + count - 1}>`;
            count = 0;
          }
        } else {
          count = 0;
          break;
        }
      }
    }

    //for top to bottom
    if (count == 0 && word[0].toUpperCase() == strArray[j][i]) {
      for (let k = 0; k < word.length; k++) {
        if (
          strArray[j + k] &&
          word[count].toUpperCase() == strArray[j + k][i]
        ) {
          count++;
          if (count == word.length) {
            indexes = `<${j},${i}>\n<${j + count - 1},${i}>`;
            count = 0;
          }
        } else {
          count = 0;
          break;
        }
      }
    }
  }
}
console.log(indexes);

File: searchForSubstring.c
/* Save the string “WELCOMETOZOHOCORPORATION” in a two dimensional array and search 
for substring like “too” in the two dimensional string both from left to right and from top to bottom.
W E L C O
M E T O Z
O H O C O
R P O R A
T I O N
And print the start and ending index as
Start index : <1,2>
End index: <3, 2> */
#include <stdio.h>
#include <string.h>
#include <math.h>

int main(){
    int n, count = 0;
    char string[100], stringArray[20][5], word[5];

    printf("Enter the length of the string: ");
    scanf("%d",&n);

    printf("Enter the string: ");
    scanf("%s",string);

    printf("Enter the word to find the index: ");
    scanf("%s",word);

    for(int i=0; i<n; i++){
      if(word[i]&&word[i]>=97) word[i]-=32;
      if(string[i]>=97) string[i]-=32;
        stringArray[i/5][i%5] = string[i];
    }

    for (int i = 0; i < n; i++) {
    if (word[0] == stringArray[i/5][(i%5)]) {
      for (int k = 0; k < strlen(word); k++) {
        if (
          stringArray[i/5][(i%5) + k] &&
          word[count] == stringArray[i/5][(i%5) + k]
        ) {
          count++;
          if (count == strlen(word)) {
            printf("<%d,%d>\n<%d,%d>\n",i/5,(i%5),i/5,(i%5)+count-1);
            count = 0;
          }
        } else {
          count = 0;
          break;
        }
      }
    }

    //for top to bottom
    if (count == 0 && word[0] == stringArray[(i%5)][i/5]) {
      for (int k = 0; k < strlen(word); k++) {
        if (
          stringArray[(i%5) + k] &&
          word[count] == stringArray[(i%5) + k][i/5]
        ) {
          count++;
          if (count == strlen(word)) {
            printf("<%d,%d>\n<%d,%d>\n",(i%5),i/5,(i%5)+count-1,i/5);
            count = 0;
          }
        } else {
          count = 0;
          break;
        }
      }
    }
}
}